data-prepper-core/src/main/java/org/opensearch/dataprepper/core/pipeline/server/DynamicPipelineUpdateUtil.java

```diff
@@ -0,0 +1,130 @@
+
+            currentProcessors = currentProcessors == null ? List.of() : currentProcessors;
+            targetProcessors = targetProcessors == null ? List.of() : targetProcessors;
+
+            // Collect single-threaded processors in current and target
+            Set<String> currentSingleThreaded = currentProcessors.stream()
+                    .map(PluginModel::getPluginName)
+                    .filter(singleThreadedProcessorNames::contains)
+                    .collect(Collectors.toSet());
+
+            Set<String> targetSingleThreaded = targetProcessors.stream()
+                    .map(PluginModel::getPluginName)
+                    .filter(singleThreadedProcessorNames::contains)
+                    .collect(Collectors.toSet());
+
+            // Only throw if target adds new single-threaded processors or modifies existing ones
+            for (String targetProcessor : targetSingleThreaded) {
+                if (!currentSingleThreaded.contains(targetProcessor)) {
+                    throw new DynamicPipelineConfigUpdateException(
+                            "Cannot add new single-threaded processor: " + targetProcessor);
+                }
+            }
+        }
+        return true;
+    }
+
+    public static void executeDynamicUpdateOfPipelineConfig(final PipelinesDataFlowModel currentPipelineDataFlowMode,
+                                                            final PipelinesDataFlowModel targetDataPipelineDataFlowMode) {
+
+
+    }
+
+    public static void checkIfSamePipelineExists(final PipelinesDataFlowModel currentPipelineDataFlowMode,
+                                                 final PipelinesDataFlowModel targetDataPipelineDataFlowMode) {
+
+        Set<String> currentPipelineNames = currentPipelineDataFlowMode.getPipelines().keySet();
+        Set<String> targetPipelineNames = targetDataPipelineDataFlowMode.getPipelines().keySet();
+
+        if (!currentPipelineNames.equals(targetPipelineNames)) {
+            String addedPipelines = targetPipelineNames.stream()
+                    .filter(name -> !currentPipelineNames.contains(name))
+                    .collect(Collectors.joining(", "));
+            String removedPipelines = currentPipelineNames.stream()
+                    .filter(name -> !targetPipelineNames.contains(name))
+                    .collect(Collectors.joining(", "));
+
+            StringBuilder errorMessage = new StringBuilder("Pipeline configuration mismatch found.");
+            if (!addedPipelines.isEmpty()) {
+                errorMessage.append(" New pipelines found: ").append(addedPipelines).append(".");
+            }
+            if (!removedPipelines.isEmpty()) {
+                errorMessage.append(" Missing pipelines: ").append(removedPipelines).append(".");
+            }
+
+            throw new DynamicPipelineConfigUpdateException(errorMessage.toString());
+        }
+    }
+
+    public static Set<String> scanForSingleThreadAnnotatedProcessorPlugins() {
```

Comment:
This is duplicating work from the `PluginFactory`. We should avoid this logic split.

I don't think the `PluginFactory` should give all `@SingleThreaded` annotations. But, maybe have a `loadPlugin` method with a `Predicate`?